crypto/s2n_hash.c

```diff
@@ -169,13 +173,270 @@ int s2n_hash_digest(struct s2n_hash_state *state, void *out, uint32_t size)
+{
+    int r;
+    unsigned int digest_size = size;
+    uint8_t expected_digest_size;
+    GUARD(s2n_hash_digest_size(state->alg, &expected_digest_size));
+    eq_check(digest_size, expected_digest_size);
+
+    switch (state->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
+        r = EVP_DigestFinal_ex(state->digest.evp.primary.ctx, out, &digest_size);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_DigestFinal_ex(state->digest.evp.primary.ctx, ((uint8_t *) out) + MD5_DIGEST_LENGTH, &digest_size);
+        r &= EVP_DigestFinal_ex(state->digest.evp.md5_secondary.ctx, out, &digest_size);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_DIGEST_FAILED);
+    }
+
     return 0;
 }
+
+static int s2n_evp_hash_copy(struct s2n_hash_state *to, struct s2n_hash_state *from)
+{
+    int r;
+    switch (from->alg) {
+    case S2N_HASH_NONE:
+        r = 1;
+        break;
+    case S2N_HASH_MD5:
+    case S2N_HASH_SHA1:
+    case S2N_HASH_SHA224:
+    case S2N_HASH_SHA256:
+    case S2N_HASH_SHA384:
+    case S2N_HASH_SHA512:
+        r = EVP_MD_CTX_copy_ex(to->digest.evp.primary.ctx, from->digest.evp.primary.ctx);
+        break;
+    case S2N_HASH_MD5_SHA1:
+        r = EVP_MD_CTX_copy_ex(to->digest.evp.primary.ctx, from->digest.evp.primary.ctx);
+        r &= EVP_MD_CTX_copy_ex(to->digest.evp.md5_secondary.ctx, from->digest.evp.md5_secondary.ctx);
+        break;
+    default:
+        S2N_ERROR(S2N_ERR_HASH_INVALID_ALGORITHM);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_COPY_FAILED);
+    }
+
+    to->alg = from->alg;
+    to->hash_impl = from->hash_impl;
+
+    return 0;
+}
+
+static int s2n_evp_hash_reset(struct s2n_hash_state *state)
+{
+    int r;
+    r = S2N_EVP_MD_CTX_RESET(state->digest.evp.primary.ctx);
+    
+    if (state->alg == S2N_HASH_MD5_SHA1) {
+        r &= S2N_EVP_MD_CTX_RESET(state->digest.evp.md5_secondary.ctx);
+    }
+
+    if (r == 0) {
+        S2N_ERROR(S2N_ERR_HASH_WIPE_FAILED);
+    }
+
+    return s2n_evp_hash_init(state, state->alg);
+}
+
+static int s2n_evp_hash_free(struct s2n_hash_state *state)
+{
+    S2N_EVP_MD_CTX_FREE(state->digest.evp.primary.ctx);
+    S2N_EVP_MD_CTX_FREE(state->digest.evp.md5_secondary.ctx);
+    state->digest.evp.primary.ctx = NULL;
+    state->digest.evp.md5_secondary.ctx = NULL;
+
+    return 0;
+}
+
+const struct s2n_hash s2n_low_level_hash = {
+    .new = &s2n_low_level_hash_new,
+    .init = &s2n_low_level_hash_init,
+    .update = &s2n_low_level_hash_update,
+    .digest = &s2n_low_level_hash_digest,
+    .copy = &s2n_low_level_hash_copy,
+    .reset = &s2n_low_level_hash_reset,
+    .free = &s2n_low_level_hash_free,
```

Comment:
Related to my previous comment - I think you could create a different struct of function pointers for each hash type that called the algorithm-specific version of the function directly instead of the s2n_low_level_hash* generic function that then calls the algorithm-specific version.